contrib/zenith/pagestore_prefetch.c

```diff
@@ -0,0 +1,401 @@
+		entry->index = curr;
+		prefetch_control->entries[curr] = entry;
+		prefetch_control->curr = (curr + 1) % prefetch_buffer_size;
+		n_prefetch_requests += 1;
+		prefetch_log("%lu: prefetch request for block %d of relation %d",
+					 GetCurrentTimestamp(), blocknum, tag.rnode.relNode);
+
+		if (prefetch_control->go_latch == NULL)
+		{
+			/* lazy start of prefetch background worker */
+			start_prefetch_worker();
+		}
+		LWLockRelease(prefetch_lock);
+		SetLatch(prefetch_control->go_latch);
+	} else
+		LWLockRelease(prefetch_lock);
+}
+
+/* Cancel merger bgwroker */
+static void
+zenith_prefetch_cancel(int sig)
+{
+	prefetch_cancel = true;
+	SetLatch(MyLatch);
+}
+
+/* Main function of prefetch bgwroker */
+void
+zenith_prefetch_main(Datum arg)
+{
+	size_t curr = 0;
+	PrefetchEntry* prefetch_entries = (PrefetchEntry*)palloc(prefetch_buffer_size*sizeof(PrefetchEntry));
+
+	pqsignal(SIGINT,  zenith_prefetch_cancel);
+	pqsignal(SIGQUIT, zenith_prefetch_cancel);
+	pqsignal(SIGTERM, zenith_prefetch_cancel);
+
+	/* We're now ready to receive signals */
+	BackgroundWorkerUnblockSignals();
+
+	prefetch_control->go_latch = MyLatch;
+
+	while (!prefetch_cancel)
+	{
+		size_t from, till;
+		size_t n_prefetched = 0;
+		(void)WaitLatch(MyLatch, WL_LATCH_SET | WL_EXIT_ON_PM_DEATH, -1L, PG_WAIT_EXTENSION);
+		ResetLatch(MyLatch);
+
+		LWLockAcquire(prefetch_lock, LW_EXCLUSIVE);
+		till = prefetch_control->curr;
+		for (from = curr; !prefetch_cancel && from != till;  from = (from + 1) % prefetch_buffer_size)
+		{
+			PrefetchEntry* entry = prefetch_control->entries[from];
+
+			if (entry == NULL) /* prefetch request was withdrawn */
+			{
+				prefetch_entries[from].state = CANCELED;
+				continue;
+			}
+			Assert(entry->state == QUEUED);
+			entry->state = IN_PROGRESS;
+			prefetch_entries[from] = *entry;
+
+			prefetch_log("%lu: send prefetch request for block %d of relation %d",
+						 GetCurrentTimestamp(), entry->tag.blockNum, entry->tag.rnode.relNode);
+
+			page_server->send((ZenithRequest) {
+				.tag = T_ZenithReadRequest,
+				.page_key = {
+					 .rnode = entry->tag.rnode,
+					 .forknum = entry->tag.forkNum,
+					 .blkno = entry->tag.blockNum
+				},
+				.lsn = entry->lsn
+		    });
+			n_prefetched += 1;
+		}
+
+		if (n_prefetched)
+		{
+			page_server->flush();
+			if (n_prefetched > max_merged_prefetch_requests)
+				max_merged_prefetch_requests = n_prefetched;
+		}
+
+		for (from = curr; !prefetch_cancel && from != till;  from = (from + 1) % prefetch_buffer_size)
+		{
+			ZenithResponse *resp;
+			PrefetchEntry* entry;
+
+			if (prefetch_entries[from].state == CANCELED)
+				continue;
+
+			/* Release lock to load buffer */
+			LWLockRelease(prefetch_lock);
+			resp = page_server->receive();
+			LWLockAcquire(prefetch_lock, LW_EXCLUSIVE);
+
+			entry = prefetch_control->entries[from];
+			if (entry != NULL
+				&& entry->lsn == prefetch_entries[from].lsn
+				&& memcmp(&entry->tag, &prefetch_entries[from].tag, sizeof(entry->tag)) == 0)
```

Comment:
We're lucky that `BufferTag` doesn't have any padding. Ideally, I'd extract this into a cmp function with a bunch of static assertions for the struct & field sizes, but whatever.

Reply:
Oh, looks like there's `BUFFERTAGS_EQUAL`.